Refactor plotting in Muon and Frequency domain GUIs

StephenSmith25 opened the pull request:
**Description of work.**
This PR refactors and reorganises the plotting code within the Muon Analysis and Frequencey Domain GUIS. The primary goals of this PR was to:

- Reduce duplication in the plotting code (as raised in #28406)
- Create a manageable code-base for plotting in the GUI as previously all the code was condensed into three files.
- Reduce dependence on a specific implementation of plotting by coding to an interface. This is likely to be beneficial as it is envisaged that frequency domain analysis will required some of its own plotting logic. 
- Base the view in the tab off a UI file so it can be easily changed. 
- As a side effect of restructuring, fixes multiple fits being plotted #28502

Left to do

- [x]  Plotting of fits (both single and sequential)
- [x] Connect QuickEditWidget with plot toolbar
- [x] Unit tests


<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
Open the Muon Analysis 2 GUI

1. Load some data e.g MUSR62260
2. Check the following functionality works (in no particular order):
a. Changing plot type
b. External plotting
c. Tiled plotting (by groups and Runs (load some more data if necessary)
d. Grouping tab (add and remove groups)
e. Rebin (options are changed in the home tab) 
f. Plot raw should switch between rebinned and raw data
g. The functionality in #28406 should be reproducible. 


<!-- Instructions for testing. -->

Fixes #28502. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

AnthonyLim23 commented:
can we connect this with issue #28502

AnthonyLim23 commented on docs/source/release/v5.1.0/muon.rst:
```diff
@@ -23,5 +23,11 @@ New Features
+Improvements
+-------------
+- Improved the plotting code the Muon Analysis interface, with a significant amount of redundant code removed.
+  This code improvement will lead to more maintainable code, which will be easier to makes changes to in the future.
```
```suggestion
  This code improvement will lead to more maintainable code, which will be more stable and easier to add new functionality.
```

AnthonyLim23 commented on scripts/Muon/GUI/Common/ADSHandler/workspace_naming.py:
```diff
@@ -96,7 +97,7 @@ def get_pair_data_workspace_name(context, pair_name, run, rebin):
 
     if rebin:
-        name += ' Rebin;'
+        name += "".join([' ', REBIN_STR, ';'])
```
I really like this change

AnthonyLim23 commented on scripts/Muon/GUI/Common/contexts/fitting_context.py:
```diff
@@ -431,9 +429,12 @@ def clear(self):
+        if self.fit_list:
+            removed_fit = self.fit_list[-1]
+            self.fit_list = self.fit_list[:-1]
+            self._number_of_fits += -1
```
```suggestion
            self._number_of_fits -= 1
```

Mantid-Matthew commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_model.py:
```diff
@@ -82,3 +125,43 @@ def get_time_workspaces_to_plot(self, current_group_pair, is_raw, plot_type):
         except AttributeError:
             return []
+
+    def get_fit_workspace_and_indices(self, fit):
```
A minor point but if we have functions like this which don't use self we can make them static. Or even a free function rather than a method. I think this is a nice thing to do because if the method doesn't rely on the state of the object (which is a really good thing) this can be useful to flag for people in the future. Both so they know they can use the method without worrying about state and so they are more cautious of adding state into the method.

StephenSmith25 replied:
Yeah good point, I've now converted these and other methods to be static.

Mantid-Matthew commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:
```diff
@@ -0,0 +1,269 @@
+class PlotWidgetPresenterCommon(HomeTabSubWidget):
+
+    def __init__(self, view: PlotWidgetViewInterface, model: PlotWidgetModel, context,
+                 figure_presenter: PlottingCanvasPresenterInterface,
```
:+1: for types **and** interfaces

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:
```diff
@@ -0,0 +1,269 @@
+        self._figure_presenter.plot_workspaces(workspace_list, indices, hold_on=hold_on, autoscale=autoscale)
+
+    def _check_if_counts_and_groups_selected(self):
+        if len(self.context.group_pair_context.selected_pairs) != 0 and \
```
could we just have `>0`?

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_view_interface.py:
```diff
@@ -0,0 +1,128 @@
+    pass
+
+
+class PlotWidgetViewInterface(metaclass=PlottingWidgetViewMeta):
```
in C++ we would call it `IPlotWidgetView`, I do like writing interface (took me ages to realise why we had the `I` at the start of the names.

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_model.py:
```diff
@@ -0,0 +1,150 @@
+        label = ''
+        fit_function_name = get_fit_function_name_from_workspace(workspace_name)
+        if fit_function_name:
+            if index == 1 or index == 3:
```
we could put `index in [1,3]`

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_presenter.py:
```diff
@@ -0,0 +1,211 @@
+    def _get_selected_subplots_from_quick_edit_widget(self):
+        subplots = self._options_presenter.get_selection()
+        if len(subplots) > 0:
+            selected_subplots = [ix - 1 for ix in list(map(int, subplots))]
```
what is `ix`?

StephenSmith25 replied:
Hmm, just index, which is just the integer representation of the subplot which are labeled ("1", "2" ect..).

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_view.py:
```diff
@@ -0,0 +1,258 @@
+            try:
+                workspace = AnalysisDataService.Instance().retrieve(workspace_name)
+            except RuntimeError:
+                continue
```
could this be a return?

StephenSmith25 replied:
Not in this case, as its plotting a list of workspaces, this just means this workspace was invalid, but the others may not be. So the for loop needs to continue.

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_view.py:
```diff
@@ -0,0 +1,258 @@
+        if axis_number >= len(self.fig.axes):
+            return
+        axis = self.fig.axes[axis_number]
+        bottom, top, = self._get_y_axis_autoscale_limts(axis)
```
why not y_max and y_min?

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plotting_canvas/plotting_canvas_view.py:
```diff
@@ -0,0 +1,258 @@
+
+    @staticmethod
+    def _get_y_axis_autoscale_limts(axis):
+        bottom = 1e9
```
why is top>bottom?

StephenSmith25 replied:
This is just the initial guess of the axis limits, having the top as -1e9 means it will be replaced by the first calculated top limit (top_i).

Mantid-Matthew commented:
Ran into a slight oddity when switching to a counts plot with a pair selected or trying to add the plotting of a pair when plotting counts. I got the following stacktrace and then switching between asymmetry and counts no longer seemed to have any effect on the plots until I added and removed some groups

```

Mantid-Matthew commented:
Another issue which I don't think is related to this PR is that when I do a simultaneous fit over several groups the simultaneous fit is stored under the name `MUSR62260; Group; top; Asymmetry; MA+ ...; Fitted; ExpDecayOsc` in the ADS which seems a bit confusing as it is actually a fit over 4 groups rather than just top. But don't think this is an issue to be fixed here.

Mantid-Matthew commented:
Two other minor things:

If I delete some of the plotted workspaces from the ADS the plots are not removed immediately but instead update once some other change is made to the plotting. Not a massive issue though as handles is cleanly in the end.

If I load some data and do a simultaneous fit over groups and then switch back and forth between the home and fitting tabs a couple of times I get the following error in the log window. Everything still seems to work though so not a massive issue.

AnthonyLim23 commented:
> Another issue which I don't think is related to this PR is that when I do a simultaneous fit over several groups the simultaneous fit is stored under the name `MUSR62260; Group; top; Asymmetry; MA+ ...; Fitted; ExpDecayOsc` in the ADS which seems a bit confusing as it is actually a fit over 4 groups rather than just top. But don't think this is an issue to be fixed here.

@Mantid-Matthew could you make an issue for this?

AnthonyLim23 commented:
the code looks good to me, just a few minor things. I still need to test it

AnthonyLim23 commented:
In FDA:
1. Load data
2. Do an FFT with the standard set up
3. select FFT all for the plot
4 Tick tiled plot
5 Change the workspace to be a different group
6. Calculate the FFT
7. It doesn't update the plot until you go to a different tab

StephenSmith25 commented:
> 
> 
> Another issue which I don't think is related to this PR is that when I do a simultaneous fit over several groups the simultaneous fit is stored under the name `MUSR62260; Group; top; Asymmetry; MA+ ...; Fitted; ExpDecayOsc` in the ADS which seems a bit confusing as it is actually a fit over 4 groups rather than just top. But don't think this is an issue to be fixed here.

I'll have to look into this more in the fitting refactor #28531

StephenSmith25 commented:
> 
> 
> Ran into a slight oddity when switching to a counts plot with a pair selected or trying to add the plotting of a pair when plotting counts. I got the following stacktrace and then switching between asymmetry and counts no longer seemed to have any effect on the plots until I added and removed some groups

This appears to be a Qt4 vs Qt5 issue. I've changed the method set_plot_type such that it is agnostic to Qt version, so hopefully this error should no longer be encountered.

StephenSmith25 commented:
> 
> 
> Two other minor things:
> 
> If I delete some of the plotted workspaces from the ADS the plots are not removed immediately but instead update once some other change is made to the plotting. Not a massive issue though as handles is cleanly in the end.
> 
> If I load some data and do a simultaneous fit over groups and then switch back and forth between the home and fitting tabs a couple of times I get the following error in the log window. Everything still seems to work though so not a massive issue.
> 
> ```
> Exception ignored in: <function MuonContext.__del__ at 0x000001880510B820>
> Traceback (most recent call last):
>  File "C:\Users\tgz88916\Documents\Mantid2\mantid\scripts\Muon\GUI\Common\contexts\muon_context.py", line 51, in __del__
>  self.ads_observer.unsubscribe()
> AttributeError: 'NoneType' object has no attribute 'unsubscribe'
> Exception ignored in: <function MuonContext.__del__ at 0x000001880510B820>
> Traceback (most recent call last):
>  File "C:\Users\tgz88916\Documents\Mantid2\mantid\scripts\Muon\GUI\Common\contexts\muon_context.py", line 51, in __del__
>  self.ads_observer.unsubscribe()
> AttributeError: 'NoneType' object has no attribute 'unsubscribe'
> ```

Thank you for finding these. The first issue occured as I wasn't redrawing the figure, this is now fixed. Not sure about the second issue if I'm honest.

StephenSmith25 commented:
> 
> 
> In FDA:
> 
>     1. Load data
> 
>     2. Do an FFT with the standard set up
> 
>     3. select FFT all for the plot
>        4 Tick tiled plot
>        5 Change the workspace to be a different group
> 
>     4. Calculate the FFT
> 
>     5. It doesn't update the plot until you go to a different tab

The graph should now update, I had wrongly removed the observer between the transform tab and plotting widget.
